### PR TITLE
fix: Fixed event gets triggered even when option is not changed

### LINKED ISSUE #23109: [Bug]: Event gets triggered even when option is not changed
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

In a TreeSelect widget, on configuring an event for the `onOptionChange` property we see that the said event gets triggered even when we do not change the selected option.

### Steps To Reproduce

1. D&D a TreeSelect widget.
2. Configure the `onOptionChange` event in such a way that we get to see an alert once we change the selected option.
3. Now open the dropdown of the widget & click on the already selected option.

**Expected:** The `onOptionChange` event should not get triggered as we did not change the option.
**Actual:** The event gets triggered.

### Public Sample App

_No response_

### Environment

Production

### Issue video log

https://www.loom.com/share/48d0980974e646e9b08af14840755430

### Version

Cloud

### PR DESCRIPTION
**Pull request by harshitpandey0426**
#### PR fixes following issue(s)
Fixes #23109

#### Type of change
Bug fix (non-breaking change which fixes an issue)

- [x] Manual
- [ ] Jest
- [x] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx [hunk 1]
```diff
@@ -582,15 +582,20 @@ class SingleSelectTreeWidget extends BaseWidget<
...
-
-    this.props.updateWidgetMetaProperty("selectedOption", value);
-    this.props.updateWidgetMetaProperty("selectedLabel", labelList?.[0] ?? "", {
-      triggerPropertyName: "onOptionChange",
-      dynamicString: this.props.onOptionChange,
-      event: {
-        type: EventType.ON_OPTION_CHANGE,
-      },
-    });
+    if (this.props.selectedOptionValue !== value) {
```
Comment (harshitpandey0426) [inline app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx:585]: This is the only change, other changes are done due to prettier formatting for code to compile

Comment (vsvamsi1) [reply app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx:585]: Hey @Ashit should the dirty assignment be moved into the value change block in line 585....just like how the select Widget behaves [Select Widget code block](https://github.com/appsmithorg/appsmith/blob/release/app/client/src/widgets/SelectWidget/widget/index.tsx#L673) 
@harshitpandey0426 Can you kindly remove this duplicate code block over here ....I just noticed it when reviewing the code of the SelectWidget https://github.com/appsmithorg/appsmith/blob/release/app/client/src/widgets/SelectWidget/widget/index.tsx#LL686C4-L688C8

Comment (harshitpandey0426) [reply app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx:585]: @vsvamsi1 removed duplicate code block

**Comment by somangshu** [comment]
This can be easily covered with a cypress test case. @harshitpandey0426 would you be willing to write it?

**Comment by harshitpandey0426** [comment]
> This can be easily covered with a cypress test case. @harshitpandey0426 would you be willing to write it?
* You mean to cover this case of onOptionChange event for SingleSelectTreeWidget with cypress?
* If yes, I can try it. Is it required in this pr?

**Comment by somangshu** [comment]
@harshitpandey0426 yes, the same pr works

File: app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx (hunk 1 above)
Comment (vsvamsi1) [reply app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx:585]: @harshitpandey0426 most of it looks good....can you also move the dirty state code into line 585 `if` block as mentioned in the earlier comment.

**Review by vsvamsi1 — CHANGES_REQUESTED** [review]

File: app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx (hunk 1 above)
Comment (harshitpandey0426) [reply app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx:582]: @vsvamsi1 moved dirty block inside if block

**Review by vsvamsi1 — APPROVED** [review]

**Comment by harshitpandey0426** [comment]
@vsvamsi1 I see couple of checks failing, is there anything pending from my end?

**Comment by somangshu** [comment]
@vsvamsi1 can you please give a sign off on this. I need to know if the failing tests are un-related and if this needs a manual QA round. Please take the required actionable on this.

**Pull request merged**